transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -311,19 +317,24 @@ private void invokeChannelRead(Object msg) {
         }
     }
 
+    private void invokeNextChannelRead(AbstractChannelHandlerContext next, Object msg) {
+        nextChannelReadInvoked = true;
+        next.invokeChannelRead(msg);
+    }
+
     @Override
     public ChannelHandlerContext fireChannelReadComplete() {
         final AbstractChannelHandlerContext next = findContextInbound();
         EventExecutor executor = next.executor();
         if (executor.inEventLoop()) {
-            next.invokeChannelReadComplete();
+            invokeNextChannelReadComplete(next);
         } else {
             Runnable task = next.invokeChannelReadCompleteTask;
             if (task == null) {
                 next.invokeChannelReadCompleteTask = task = new Runnable() {
                     @Override
                     public void run() {
-                        next.invokeChannelReadComplete();
+                        invokeNextChannelReadComplete(next);
```

Comment:
Let's say you cached the next context's invokeChannelReadCompleteTask. The task here will call this.invokeNextChannelReadComplete() where `this` is the previous context. So far so good.

However, if a user removes the previous context from the pipeline, `next`'s previous context is not `this` anymore.  Hence, `invokeChannelReadCompleteTask` will update the flag of the wrong context, which has been removed already.